pkg/agent/packetcapture/packetcapture_controller.go

```diff
@@ -0,0 +1,774 @@
+					c.numRunningCaptures += 1
+					return packetCapturePhaseRunning, c.startPacketCapture(pcName)
+				}
+				klog.InfoS("PacketCapture running count reach limit, waiting in queue", "name", pc.Name)
+				return packetCapturePhasePending, fmt.Errorf("PacketCapture running count reach limit, waiting in queue")
+			}()
+			c.captures[pcName] = newStatus
+			return newStatus, err
+		}
+		return status, err
+	}()
+	if err := c.patchStatusIfNeeded(pcName, status); err != nil {
+		return fmt.Errorf("error when patching status: %w", err)
+	}
+	return nil
+}
+
+func (c *Controller) cleanupPacketCapture(pcName string) {
+	path := nameToPath(pcName)
+	exist, err := afero.Exists(defaultFS, path)
+	if err != nil {
+		klog.ErrorS(err, "Failed to check if path exists", "path", path)
+	}
+	if !exist {
+		return
+	}
+	if err := defaultFS.Remove(path); err == nil {
+		klog.V(2).InfoS("Deleted the captured pcap file successfully", "name", pcName, "path", path)
+	} else {
+		klog.ErrorS(err, "Failed to delete the captured pcap file", "name", pcName, "path", path)
+	}
+}
+
+func getPacketFileAndWriter(name string) (afero.File, *pcapgo.NgWriter, error) {
+	filePath := nameToPath(name)
+	var file afero.File
+	if _, err := os.Stat(filePath); err == nil {
+		klog.Warningf("the packet file %s already exists. This may be caused by an unexpected termination, will delete it", filePath)
+		if err := defaultFS.Remove(filePath); err != nil {
+			return nil, nil, err
+		}
+	}
+	file, err := defaultFS.Create(filePath)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to create pcapng file: %w", err)
+	}
+	writer, err := pcapgo.NewNgWriter(file, layers.LinkTypeEthernet)
+	if err != nil {
+		return nil, nil, fmt.Errorf("couldn't initialize a pcap writer: %w", err)
+	}
+	return file, writer, nil
+}
+
+// getTargetCaptureDevice is trying to locate the target device for packet capture. If the target Pod does not exist on the current Node,
+// the agent on this Node will not perform the capture. In the PacketCapture spec, at least one of `.Spec.Source.Pod` or `.Spec.Destination.Pod`
+// should be set.
+func (c *Controller) getTargetCaptureDevice(pc *crdv1alpha1.PacketCapture) *string {
+	var pod, ns string
+	if pc.Spec.Source.Pod != nil {
+		pod = pc.Spec.Source.Pod.Name
+		ns = pc.Spec.Source.Pod.Namespace
+	} else {
+		pod = pc.Spec.Destination.Pod.Name
+		ns = pc.Spec.Destination.Pod.Namespace
+	}
+
+	podInterfaces := c.interfaceStore.GetContainerInterfacesByPod(pod, ns)
+	if len(podInterfaces) == 0 {
+		return nil
+	}
+
+	return &podInterfaces[0].InterfaceName
+}
+
+func (c *Controller) startPacketCapture(name string) error {
+	klog.V(4).InfoS("Started processing PacketCapture", "name", name)
+	var err error
+	pcState := &packetCaptureState{name: name}
+	cleanup := func() {
+		if err != nil {
+			klog.ErrorS(err, "PackageCapture failed", "name", name)
+			err := c.updatePacketCaptureStatus(name, pcState.capturedPacketsNum, "", err)
```

Comment:
If I understand the code correctly, `pcState.capturedPacketsNum` will always be 0 here. This is because this error handling code doesn't handle errors in `performCapture`.
So the code is misleading.

Reply:
you are right, even it didn't make a different. `status. NumberCaptured` will respect the old value if the new one is 0. so after `performCapture` update the number, it's settled. will update this part.